### PR TITLE
Add DT_PARENT()

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
A few prep work commits to gen_defines.py, followed by a new DT_PARENT() API in devicetree.h.

### CODE REVIEW

File: include/devicetree.h [hunk 1]
```diff
@@ -208,6 +208,26 @@
...
+/**
+ * @brief Get a node identifier for a parent node
+ *
+ * Example devicetree fragment:
+ *
+ *     parent: parent-node {
+ *             child: child-node {
+ *                     ....
+ *             };
+ *     }
```
Comment (mbolivar-nordic) [inline include/devicetree.h:220]: I missed a semicolon here.

**Comment by de-nordic** [comment]
The _PARENT node should be defined for each node, so that it would be possible to traverse back from child note, for example from:
DT_N_S_soc_S_uart_40002000_P_interrupts
it would be nice to get
DT_N_S_soc_S_uart_40002000.
This way it may be possible, for example to get DT entry for flash that is holding given partition, e.g.:
DT_PARENT(boot0_partition)
would return partitions node, that contains boot0_partition) of some flash device, but
DT_PARENT(DT_PARENT(boot0_partition)) would give flash node.

**Comment by mbolivar-nordic** [comment]
> The _PARENT node should be defined for each node, 

It is defined for each node. That's what this PR is doing.

> so that it would be possible to traverse back from child note, for example from:
> `DT_N_S_soc_S_uart_40002000_P_interrupts`

That's a property value, not a node identifier. I'm confused. Why do you need to go from a property value to the node?

> it would be nice to get
> `DT_N_S_soc_S_uart_40002000`.

I'm not sure what you mean, but it doesn't work this way. Property macros expand to C rvalues, unlike node identifiers, which [are not values](https://docs.zephyrproject.org/latest/guides/dts/api-usage.html#node-identifiers-are-not-values).

**Comment by de-nordic** [comment]
What I mean is that having this:
```#define DT_N_NODELABEL_boot_partition DT_N_S_soc_S_flash_controller_4001e000_S_flash_0_S_partitions_S_partition_0```
I want to traverse back to this identifier:
```DT_N_S_soc_S_flash_controller_4001e000_S_flash_0```
So that I could use it to get properties from it, for example use the `DT_LABEL` to get the dev name.

**Comment by mbolivar-nordic** [comment]
> So that I could use it to get properties from it, for example use the `DT_LABEL` to get the dev name.

D'oh. That was my goal, and I failed. Fixed now, thank you.

**Comment by de-nordic** [comment]
> 
> D'oh. That was my goal, and I failed. Fixed now, thank you.

Now it works as expected, thanks!

**Review by galak — APPROVED** [review]

**Pull request merged**